### PR TITLE
blkin: remove blkin trace submodule and cleanup

### PR DESCRIPTION
**Pull request by zenomri**
This PR comes to remove the blkin trace submodule, and all of its usages
including librados API methods and tests that used it.

since the opentelemetry tracing infrastructure becomes mature in ceph, and the blkin is out of use, we can safely remove/replace it

this cleanup also changes some of the messages that were carrying that blkin trace info,
so now those messages carries the o-tel trace info instead of blkin - but for backward compatibility we still support decode of blkin struct for old messages

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/cls/rgw/cls_rgw.cc [hunk 1]
```diff
@@ -149,7 +150,8 @@ static void bi_log_index_key(cls_method_context_t hctx, string& key, string& id,
 static int log_index_operation(cls_method_context_t hctx, const cls_rgw_obj_key& obj_key,
                                RGWModifyOp op, const string& tag, real_time timestamp,
                                const rgw_bucket_entry_ver& ver, RGWPendingState state, uint64_t index_ver,
-                               string& max_marker, uint16_t bilog_flags, string *owner, string *owner_display_name, rgw_zone_set *zones_trace)
+                               string& max_marker, uint16_t bilog_flags, string *owner, string *owner_display_name, rgw_zone_set *zones_trace,
+                               const jspan_context* bilog_trace = nullptr)
```
Comment (yuvalif) [inline src/cls/rgw/cls_rgw.cc:154]: the bilog trace change seems unrelated to the blkin cleanup

File: src/common/tracer.cc [hunk 2]
```diff
@@ -59,7 +59,7 @@ jspan Tracer::add_span(opentelemetry::nostd::string_view span_name, const jspan&
 }
 
 jspan Tracer::add_span(opentelemetry::nostd::string_view span_name, const jspan_context& parent_ctx) {
-  if (is_enabled() && parent_ctx.IsValid()) {
```
Comment (yuvalif) [inline src/common/tracer.cc:62]: changes in this file seem unrelated to the blkin cleanup

File: src/librbd/ImageCtx.h [hunk 3]
```diff
@@ -19,7 +19,7 @@
 #include "common/event_socket.h"
 #include "common/Readahead.h"
 #include "common/snap_types.h"
-#include "common/zipkin_trace.h"
+#include "common/tracer.h"
```
Comment (yuvalif) [inline src/librbd/ImageCtx.h:22]: do you really need this header here?

File: src/librbd/Utils.h [hunk 4]
```diff
@@ -9,7 +9,7 @@
 #include "include/ceph_assert.h"
 #include "include/Context.h"
 #include "common/snap_types.h"
-#include "common/zipkin_trace.h"
+#include "common/tracer.h"
```
Comment (yuvalif) [inline src/librbd/Utils.h:12]: why is this needed?

File: src/librbd/crypto/CryptoObjectDispatch.cc [hunk 5]
```diff
@@ -32,7 +32,7 @@ struct C_AlignedObjectReadRequest : public Context {
     uint64_t object_no;
     io::ReadExtents* extents;
     IOContext io_context;
-    const ZTracer::Trace parent_trace;
+    const jspan_context parent_trace{false, false};
```
Comment (yuvalif) [inline src/librbd/crypto/CryptoObjectDispatch.cc:35]: why did you create the trace here, and not just passed the null trace (as in the previous examples)?

File: src/librbd/io/ImageRequest.h [hunk 6]
```diff
@@ -29,41 +29,39 @@ class ImageRequest {
 public:
   typedef std::vector<std::pair<uint64_t,uint64_t> > Extents;
 
-  virtual ~ImageRequest() {
-    m_trace.event("finish");
-  }
+  virtual ~ImageRequest() {}
```
Comment (yuvalif) [inline src/librbd/io/ImageRequest.h:32]: nit:
```suggestion
  virtual ~ImageRequest() = default;
```

File: src/librbd/object_map/UpdateRequest.h [hunk 7]
```diff
@@ -41,21 +41,18 @@ class UpdateRequest : public Request {
...
       m_ignore_enoent(ignore_enoent)
-  {
-    m_trace.event("start");
-  }
-  virtual ~UpdateRequest() {
-    m_trace.event("finish");
-  }
+  {}
+
+  virtual ~UpdateRequest() {}
```
Comment (yuvalif) [inline src/librbd/object_map/UpdateRequest.h:55]:
```suggestion
  virtual ~UpdateRequest() = default;
```

File: src/messages/MOSDECSubOpRead.h [hunk 8]
```diff
@@ -20,8 +20,8 @@
 
 class MOSDECSubOpRead : public MOSDFastDispatchOp {
 private:
-  static constexpr int HEAD_VERSION = 3;
-  static constexpr int COMPAT_VERSION = 1;
+  static constexpr int HEAD_VERSION = 4;
+  static constexpr int COMPAT_VERSION = 2;
```
Comment (yuvalif) [inline src/messages/MOSDECSubOpRead.h:24]: why did you change the `COMPAT_VERSION`?

File: src/messages/MOSDECSubOpReadReply.h [hunk 9]
```diff
@@ -20,8 +20,8 @@
 
 class MOSDECSubOpReadReply : public MOSDFastDispatchOp {
 private:
-  static constexpr int HEAD_VERSION = 2;
-  static constexpr int COMPAT_VERSION = 1;
+  static constexpr int HEAD_VERSION = 3;
+  static constexpr int COMPAT_VERSION = 2;
```
Comment (yuvalif) [inline src/messages/MOSDECSubOpReadReply.h:24]: same as: https://github.com/ceph/ceph/pull/48971/files#r1027317533

File: src/messages/MOSDECSubOpWrite.h [hunk 10]
```diff
@@ -20,8 +20,8 @@
 
 class MOSDECSubOpWrite : public MOSDFastDispatchOp {
 private:
-  static constexpr int HEAD_VERSION = 2;
-  static constexpr int COMPAT_VERSION = 1;
+  static constexpr int HEAD_VERSION = 3;
+  static constexpr int COMPAT_VERSION = 2;
```
Comment (yuvalif) [inline src/messages/MOSDECSubOpWrite.h:24]: same as: https://github.com/ceph/ceph/pull/48971/files#r1027317533

File: src/messages/MOSDECSubOpWriteReply.h [hunk 11]
```diff
@@ -20,8 +20,8 @@
 
 class MOSDECSubOpWriteReply : public MOSDFastDispatchOp {
 private:
-  static constexpr int HEAD_VERSION = 2;
-  static constexpr int COMPAT_VERSION = 1;
+  static constexpr int HEAD_VERSION = 3;
+  static constexpr int COMPAT_VERSION = 2;
```
Comment (yuvalif) [inline src/messages/MOSDECSubOpWriteReply.h:24]: same as: https://github.com/ceph/ceph/pull/48971/files#r1027317533

File: src/messages/MOSDOpReply.h [hunk 12]
```diff
@@ -31,8 +31,8 @@
 
 class MOSDOpReply final : public Message {
 private:
-  static constexpr int HEAD_VERSION = 8;
-  static constexpr int COMPAT_VERSION = 2;
+  static constexpr int HEAD_VERSION = 9;
+  static constexpr int COMPAT_VERSION = 3;
```
Comment (yuvalif) [inline src/messages/MOSDOpReply.h:35]: same as: https://github.com/ceph/ceph/pull/48971/files#r1027317533

File: src/messages/MOSDRepOp.h [hunk 13]
```diff
@@ -24,8 +24,8 @@
 
 class MOSDRepOp final : public MOSDFastDispatchOp {
 private:
-  static constexpr int HEAD_VERSION = 3;
-  static constexpr int COMPAT_VERSION = 1;
+  static constexpr int HEAD_VERSION = 4;
+  static constexpr int COMPAT_VERSION = 2;
```
Comment (yuvalif) [inline src/messages/MOSDRepOp.h:28]: same as: https://github.com/ceph/ceph/pull/48971/files#r1027317533

File: src/messages/MOSDRepOpReply.h [hunk 14]
```diff
@@ -29,8 +29,8 @@
 
 class MOSDRepOpReply final : public MOSDFastDispatchOp {
 private:
-  static constexpr int HEAD_VERSION = 2;
-  static constexpr int COMPAT_VERSION = 1;
+  static constexpr int HEAD_VERSION = 3;
+  static constexpr int COMPAT_VERSION = 2;
```
Comment (yuvalif) [inline src/messages/MOSDRepOpReply.h:33]: same as: https://github.com/ceph/ceph/pull/48971/files#r1027317533

File: src/rgw/rgw_cr_rados.cc [hunk 15]
```diff
@@ -722,7 +722,8 @@ int RGWAsyncFetchRemoteObj::_send_request(const DoutPrefixProvider *dpp)
   rgw::sal::RadosObject src_obj(store, key, &bucket);
   rgw::sal::RadosBucket dest_bucket(store, dest_bucket_info);
   rgw::sal::RadosObject dest_obj(store, dest_key.value_or(key), &dest_bucket);
-    
+  dest_obj.set_trace(std::move(trace_ctx));
```
Comment (yuvalif) [inline src/rgw/rgw_cr_rados.cc:725]: is this part of the cleanup?

File: src/rgw/rgw_data_sync.cc [hunk 16]
```diff
@@ -4731,6 +4741,8 @@ int RGWBucketShardIncrementalSyncCR::operate(const DoutPrefixProvider *dpp)
           marker_tracker.try_update_high_marker(cur_id, 0, entry->timestamp);
           continue;
         }
+        trace = tracing::rgw::tracer.add_span("RGWBucketShardIncrementalSyncCR::operate", entry->bi_trace);
```
Comment (yuvalif) [inline src/rgw/rgw_data_sync.cc:4744]: seems unrelated

**Comment by yuvalif** [comment]
@zenomri  please remove  unrelated commits (e.g. bi log)

File: src/cls/rgw/cls_rgw.cc (hunk 1 above)
Comment (zenomri) [reply src/cls/rgw/cls_rgw.cc:154]: Indeed. It's part of the previous PR, which we depend on.

File: src/messages/MOSDECSubOpRead.h (hunk 8 above)
Comment (zenomri) [reply src/messages/MOSDECSubOpRead.h:24]: because the change to the message structure is not just adding something at the end of the msg. according to https://docs.ceph.com/en/latest/dev/encoding/

File: src/cls/rgw/cls_rgw.cc (hunk 1 above)
Comment (yuvalif) [reply src/cls/rgw/cls_rgw.cc:154]: added DNM tag, as this PR will have to wait until thew other is merged.

File: src/messages/MOSDECSubOpRead.h (hunk 8 above)
Comment (cbodley) [reply src/messages/MOSDECSubOpRead.h:24]: you can't just bump compat_version by 1, you'd have to set it to the minimum version you can still decode. if you remove stuff in the v4 encoding, then compat_version also needs to be 4. but that means we'd break all backward compatibility with existing clients

so we need to do this without changing compat versions. unfortunately i think that means we still need to call `encode_trace()`/`decode_trace()` even if we're going to ignores its values

Comment (yuvalif) [reply src/messages/MOSDECSubOpRead.h:24]: i dont think we need to change the compatibility version at all. we kept the blkin info struct and we are decoding it

Comment (zenomri) [reply src/messages/MOSDECSubOpRead.h:24]: thanks for the clarification! will fix it

Comment (cbodley) [reply src/messages/MOSDECSubOpRead.h:24]: > i dont think we need to change the compatibility version at all. we kept the blkin info struct and we are decoding it

if we remove the `encode_trace()` part from the v4 encoding, then receivers still running v3 would fail to decode

Comment (yuvalif) [reply src/messages/MOSDECSubOpRead.h:24]: ok. so, should we just keep that structure encoded as well?

**Comment by yuvalif** [comment]
@zenomri could you also remove [blkin docs](https://docs.ceph.com/en/latest/dev/blkin/) ?

**Comment by rzarzynski** [comment]
From core PR scrub: what's the status of this PR? It's severely conflicted. Should we close?

**Comment by yuvalif** [comment]
> From core PR scrub: what's the status of this PR? It's severely conflicted. Should we close?

it is a followup PR to: https://github.com/ceph/ceph/pull/52114
once the above is merged, i'll be able to finis this one

**Comment by yuvalif** [comment]
lets keep it opened for now

**Comment by yuvalif** [comment]
unstale please

**Comment by yuvalif** [comment]
unstale please

**Comment by cbodley** [comment]
would be great to land this and https://github.com/ceph/ceph/pull/52114 for squid!

**Comment by yuvalif** [comment]
unstale please

**Comment by adamemerson** [comment]
@yuvalif Is this something we still need? I thought the new tracing obsoleted blkn.

**Comment by adamemerson** [comment]
Wait, never mind, this is removing blkn.

**Comment by cbodley** [comment]
https://github.com/ceph/ceph/pull/55499 added this release note for squid:
> The blkin tracing feature is now deprecated in favor of Opentracing and will be removed in a later release.

i'd support removing this for T if someone's willing to work through the rebase

**Comment by adamemerson** [comment]
Closing in favor of https://github.com/ceph/ceph/pull/59365 , but thank you for getting everything started and doing so much work!
